### PR TITLE
Jit Generators and asyncs + minor enhancements

### PR DESCRIPTION
**Pull request by rhuanjl**
Enable the JIT for Generators and Async functions (including Async Generators).

This builds on work done by @zenparsing and @nhat-nguyen to finally enable jitting functions containing yield or await.

I've also included a couple of minor improvements that I'd written a while ago and been waiting to merge:
1. cache the await object - this was being allocated and freed for every `await` it's now allocated at the top of an async function and freed at the end.
1. cache async generator continuation methods - these were being allocated and freed for every yield or await inside an async generator, now allocated at the start of the function and freed at the end.

@ppenzin not a significant diff by line count but as this enabling a lot of other code that has not been enabled before may need some careful thought.

Fixes: #5877

### CODE REVIEW

**Comment by ppenzin** [comment]
I am not immediately able to reproduce it from the CI log, re-running the test suite on my machine. Hopefully will have a better report on this in a bit.

**Comment by nhat-nguyen** [comment]
> The lone test failure is very strange - it's hitting an assertion using time travel debugging and the `NoNative` flag, it does not occur if you instead use the `-off:backend` flag which is functionally equivalent to `NoNative` so clearly there is something odd happening here.
> 
> Also weird that a change to enable jitting things not previously jitted would hit an assertion when using a flag that disables the jit.

I could be wrong, but if my memory serves me correctly, -off:backend doesn't turn off the jit completely. So -nonative and -off:backend aren't exactly the same. -nonative is interpreter only.

**Comment by rhuanjl** [comment]
I had a bit of a dig - as far as I can see the issue is that TTD doesn't have explicit support for an await object. It still restores the right information but doesn't set the type. Currently the only impact of this is hitting that assertion the code runs to completion correctly with the assertion removed.

I'm unsure why playing around with jit/nojit mode hits this but it might be to do with the point in the code that the snapshot is taken for the TTD.

AsyncGenerators - which would be impacted more significantly by this - do not support TTD at all, but I'm not planning on fixing that for now hence I've added a change to stop the assertion firing when in TTDReplay mode.

**Comment by rhuanjl** [comment]
Tests failed again due to test machine problems - 95% of the way through the windows 10 tests it then suddenly couldn't find the disk.

Going back over the code I found a bug that wasn't being tested:
Generator.prototype.next.call(asyncGenerator) should throw a TypeError but was running to completion - in the context of jitted AsyncGenerators that could easily produce undefined behaviour so adding a fix for that.

Also noticed that AsyncGenerator.prototype.next.call(invalid) was throwing the slightly odd error message: `AsyncGenerator.prototype.next 'this' is not a %s object` absolutely minor but adding a fix for that at the same time - as well as tests of that code path as it was untested.

File: lib/Runtime/Library/JavascriptGenerator.cpp [hunk 1]
```diff
@@ -231,7 +231,7 @@ Var JavascriptGenerator::EntryNext(RecyclableObject* function, CallInfo callInfo
 
     AUTO_TAG_NATIVE_LIBRARY_ENTRY(function, callInfo, _u("Generator.prototype.next"));
 
-    if (!VarIs<JavascriptGenerator>(args[0]))
+    if (!VarIs<DynamicObject>(args[0]) || JavascriptOperators::GetTypeId(args[0]) != TypeIds_Generator)
```
Comment (rhuanjl) [inline lib/Runtime/Library/JavascriptGenerator.cpp:234]: VarIs<JavascriptGenerator> returns true for AsyncGenerators (which are a subclass of Generators) hence had to change these conditions to use an explicit check for Generators instead.

**Comment by rhuanjl** [comment]
Planning to merge this when the CI stops falling over.

**Comment by rhuanjl** [comment]
So due to the way the switch was setup my original PR wasn't actually enabling the JIT  - having fixed that I've uncovered 2 significant bugs - I've fixed one but still trying to track through the second.

Currently if a generator function hits a BailoutOnNoProfile it restores with the wrong variable values - which results in inconsistent observable behaviour, closing this until I can fix this issue.

**Comment by rhuanjl** [comment]
Well, two bugs down and found another one.

This one seems to behave correctly BUT hits an assertion in debug builds, will hopefully find a fix shortly.

**Comment by rhuanjl** [comment]
I'll try and have another look at this in the next two weeks - really want to get it working BUT ugh it is not simple...

**Comment by rhuanjl** [comment]
Closing this - I've moved the parts that are ready to land into #6533 going to keep working on the other parts.

**Pull request closed without merging**